certdb/certdb.go

```
@@ -0,0 +1,227 @@
+func wrapCertStoreError(err error) error {
+	if err != nil {
+		return cferr.Wrap(cferr.CertStoreError, cferr.Unknown, err)
+	}
+	return nil
+}
+
+// InsertCertificate puts a CertificateRecord into db.
+func InsertCertificate(db *sql.DB, cr *CertificateRecord) error {
+	res, err := db.Exec(
+		insertSQL,
+		cr.Serial,
+		cr.CALabel,
+		cr.Status,
+		cr.Reason,
+		cr.Expiry.UTC(),
+		cr.RevokedAt,
+		cr.PEM,
+	)
+	if err != nil {
+		return wrapCertStoreError(err)
+	}
+
+	numRowsAffected, _ := res.RowsAffected()
+
+	if numRowsAffected != 1 {
+		return wrapCertStoreError(fmt.Errorf("%d rows are affected, should be 1 row", numRowsAffected))
+	}
+
+	return nil
+}
+
+// GetCertificate gets a CertificateRecord indexed by serial.
+func GetCertificate(db *sql.DB, serial string) (*CertificateRecord, error) {
+	cr := new(CertificateRecord)
+	rows, err := db.Query(fmt.Sprintf(selectSQL, sqlstruct.Columns(*cr)), serial)
+	if err != nil {
+		return nil, wrapCertStoreError(err)
+	}
+	defer rows.Close()
+
+	if rows.Next() {
+		return cr, wrapCertStoreError(sqlstruct.Scan(cr, rows))
+	}
+	return nil, nil
+}
+
+// GetUnexpiredCertificates gets all unexpired certificate from db.
+func GetUnexpiredCertificates(db *sql.DB) (crs []*CertificateRecord, err error) {
+	cr := new(CertificateRecord)
+	rows, err := db.Query(fmt.Sprintf(selectAllUnexpiredSQL, sqlstruct.Columns(*cr)))
+	if err != nil {
+		return nil, wrapCertStoreError(err)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		err = sqlstruct.Scan(cr, rows)
+		if err != nil {
+			return nil, wrapCertStoreError(err)
+		}
+		crs = append(crs, cr)
+	}
+
+	return crs, nil
+}
+
+// RevokeCertificate updates a certificate with a given serial number and marks it revoked.
+func RevokeCertificate(db *sql.DB, serial string, reasonCode int) error {
+	result, err := db.Exec(updateRevokeSQL, reasonCode, serial)
+
+	if err != nil {
+		return wrapCertStoreError(err)
+	}
+
+	numRowsAffected, _ := result.RowsAffected()
+
+	if numRowsAffected != 1 {
```

Should we return a different error if a certificate with a given serial isn't in the database? That's a much less severe and more useful error - we'd want to return that all the way up to the client.